### PR TITLE
list updated plugins upon completion of bin/plugin update

### LINKED ISSUE #2625: bin/plugin update should list updated plugins upon completion
it currently silently updates plugins but does not show what was or was not updated. 

to support this we should gather the plugin specs prior and after the update, compare and report.

### PR DESCRIPTION
**Pull request by colinsurprenant**
fixes #2625 
with this the new behaviour is:
- any local plugin gem can be updated, not just the one in the Gemfile, so any plugin showing in `bin/plugin list` can be updated.
- all updated plugins will display:
```
Updated logstash-input-awesome 0.0.1 to 1.2.3
```
- if the update process triggers the installation of a new plugin, it will show:
```
Installed logstash-codec-awesome 1.2.3
```
- otherwise if nothing gets installed or updated it will show:
```
No plugin updated
```

### CODE REVIEW

**Comment by suyograo** [comment]
Testing this

File: lib/logstash/pluginmanager/update.rb [hunk 1]
```diff
@@ -16,18 +16,20 @@ def execute
...
     original_gemset = gemfile.gemset.copy
 
+    previous_gem_specs_map = find_latest_gem_specs
+
     # create list of plugins to update
     plugins = unless plugin_list.empty?
-      not_installed = plugin_list.find{|plugin| !LogStash::PluginManager.installed_plugin?(plugin, gemfile)}
-      raise(LogStash::PluginManager::Error, "Plugin #{not_installed} has not been previously installed, aborting") if not_installed
+      not_installed = plugin_list.select{|plugin| !previous_gem_specs_map.has_key?(plugin.downcase)}
+      raise(LogStash::PluginManager::Error, "Plugin #{not_installed.join(', ')} has not been previously installed, aborting") unless not_installed.empty?
```
Comment (jordansissel) [inline lib/logstash/pluginmanager/update.rb]: > has not been previously installed, aborting

how about: "is not installed, so I cannot upgrade it."

File: lib/logstash/pluginmanager/update.rb [hunk 2]
```diff
@@ -43,6 +45,33 @@ def execute
...
+        if spec.version != previous_gem_specs_map[name].version
+          puts("Updated #{spec.name} #{previous_gem_specs_map[name].version.to_s} to #{spec.version.to_s}")
+          update_count += 1
+        end
+      else
+        puts("Installed #{spec.name} #{spec.version.to_s}")
+        update_count += 1
+      end
+    end
+    puts("No plugin updated") if update_count.zero?
```
Comment (jordansissel) [inline lib/logstash/pluginmanager/update.rb]: `.zero?` is ... weird.

Comment (colinsurprenant) [inline lib/logstash/pluginmanager/update.rb]:
```ruby
irb(main):001:0> 0.zero?
=> true
irb(main):002:0> 1.zero?
=> false
```

bad weird or weird weird? ;)

File: lib/logstash/pluginmanager/update.rb (hunk 1 above)
Comment (colinsurprenant) [inline lib/logstash/pluginmanager/update.rb]: updated message

**Comment by suyograo** [comment]
Works: 
```
suyog@machine:~/ws/logstash (pr/2739)$ bin/plugin update logstash-output-elasticsearch
Updating logstash-output-elasticsearch
Updated logstash-output-elasticsearch 0.1.17 to 0.1.18
```

**Comment by suyograo** [comment]
Running again shows:
```
suyog@machine:~/ws/logstash (pr/2739)$ bin/plugin update logstash-output-elasticsearch
Updating logstash-output-elasticsearch
No plugin updated
```

**Comment by suyograo** [comment]
:shipit:

**Pull request closed without merging**